odiglet/pkg/ebpf/configprovider.go

```diff
@@ -0,0 +1,69 @@
+package ebpf
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/odigos-io/odigos/odiglet/pkg/log"
+)
+
+type ConfigProvider[C any] struct {
+	configChan    chan C
+	initialConfig C
+
+	stoppedMutex sync.Mutex
+	stopped      bool
+}
+
+// NewConfigProvider creates a new configProvider with the given initial config.
+// It allows for updating the configuration of a running instrumentation.
+func NewConfigProvider[C any](initialConfig C) *ConfigProvider[C] {
+	return &ConfigProvider[C]{
+		initialConfig: initialConfig,
+		configChan:    make(chan C),
+	}
+}
+
+func (c *ConfigProvider[C]) InitialConfig(_ context.Context) C {
+	return c.initialConfig
+}
+
+func (c *ConfigProvider[C]) Shutdown(_ context.Context) error {
+	c.stoppedMutex.Lock()
+	defer c.stoppedMutex.Unlock()
+
+	if c.stopped {
+		return nil
+	}
+
+	close(c.configChan)
+	c.stopped = true
+	return nil
+}
+
+func (c *ConfigProvider[C]) Watch() <-chan C {
+	return c.configChan
+}
+
+func (c *ConfigProvider[C]) SendConfig(ctx context.Context, newConfig C) error {
+	c.stoppedMutex.Lock()
+	defer c.stoppedMutex.Unlock()
+
+	if c.stopped {
+		log.Logger.Info("SendConfig called on stopped configProvider, the supplied config will be ignored")
+		return nil
+	}
+
+	// send a config or potentially return an error on timeout
+	// TODO: we should decide on a timeout value for sending config.
+	select {
+	case c.configChan <- newConfig:
+		return nil
+	case <-ctx.Done():
```

Comment:
We assume the caller passed `ctx` which has a timeout, maybe a better design would be to get both context and `time.Time` or something as arguments and wrap the context in this function instead of assuming the caller did it